### ISSUE TITLE
WAF rate limits: sign in routes

### ISSUE THREAD

**Issue by sastels** [cds-snc/notification-planning#799]
# Description

As a Notify stakeholder,
I want to reduce the attacks on Notify, 
so that our software is less discoverable with vulnerabilities
and only allow legit traffic to the web application.

## WHY are we building?

Rate limit attacks on the sign in urls:
/sign-in, /register. /forgot-password, /forced-password-reset

## WHAT are we building?

Add additional rules in our AWS WAF firewall to rate limit traffic from same IP and other factors.

## VALUE created by our solution

Less discoverable, less support time spent on attacks.

# Acceptance Criteria** (Definition of done) 

- [x] Check our WAF data to see what the historical numbers for these endpoints are
- [x] Repeated traffic from same IP to these endpoints should be rate limited.

# QA Steps

Note: deployed to production.

- [ ] Send many requests to /sign-in and wait to get blocked.
- [ ] While blocked, verify that you can still go to the main page
- [ ] While blocked, verify that you can still send notifications through the api
- [ ] Wait 5 minutes and verify that you are no longer blocked.

Script to send lots of requests:
```
#!/bin/bash
for ((i=0; i<$1; i++))
do
printf "\n${i}: Status "
curl -s -o /dev/null -w "%{http_code}" https://$2
sleep 1
done
echo ""
```
Usage: 
```
./repeat_curl.sh 200 "notification.canada.ca/sign-in"
```

**Comment by jzbahrai** [comment]
The is only on staging not on prd yet. @jimleroyer and @sastels will talk about further QA steps and deployment.

**Comment by yaelberger-commits** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/799?planning-poker) with ZenHub @andrewleith @jimleroyer @jzbahrai @sastels

**Comment by sastels** [comment]
Merged into prod now.

Script to send lots of requests:

#!/bin/bash
for ((i=0; i<$1; i++))
do
    printf "\n${i}: Status "
    curl -s -o /dev/null -w "%{http_code}" https://$2
    sleep 1
done
echo "" 

Usage: ./repeat_curl.sh 200 "notification.canada.ca/sign-in"

**Comment by jzbahrai** [comment]
This is in production - someone can't sign in more than a 100 times in prod

**Comment by jimleroyer** [comment]
Tested with a Locust Python test. I hit protected endpoints such as `/sign-in`, `/register`, `/forgot-password`, 

The WAF rule specifies 100 hits within 5 minutes as a threshold but that rule is evaluated every 30 seconds. It's important to note because that means that someone can hit as many times within that 30 period time frame, i.e. it's not when the same IP hits 100 times that it blocks automatically afterward, it's when the same IP hits more than 100 times when the 30 seconds evaluation period kicks in.

For example, in my tests, it took around 30-45 seconds for the rule to take effect generally where I could squeeze in 3,515 successful hits within 30 seconds against the URLs before getting blocked.

In the test below, we can observe a moderate 444 successful hits within a 27 seconds time period before the failures kick in. 

![image.png](https://images.zenhubusercontent.com/614b3e364039e00aa3137935/f4987dd6-49d9-47ae-80f0-5cc8ed73b87f)

![image.png](https://images.zenhubusercontent.com/614b3e364039e00aa3137935/09ee5042-8ddc-4067-b6b8-88df74270d86)

The screen I got in Chrome while I was blocked on one of the URL:
![image.png](https://images.zenhubusercontent.com/614b3e364039e00aa3137935/1080f879-a85e-49e6-9b9b-3e863646bd17)

When the block is activated, all listed URLs are unreachable. Other pages such as the homepage works correctly through (as expected).

The WAF console correctly reports the blocked requests once the rule activated:
![image.png](https://images.zenhubusercontent.com/614b3e364039e00aa3137935/1e39f8ea-5db9-4afb-9ddc-b6ea4409c425)

I was unblocked once the 5 minutes period absorbed all the hits I sent during the tests.

**Comment by jimleroyer** [comment]
Added automated Locust test in the `notification-api/tests-perf` folder:
https://github.com/cds-snc/notification-api/pull/1580/files

**Comment by jimleroyer** [comment]
This task has been QA'ed and can be considered done. 👍

**Comment by sastels** [comment]
yeah, agreed the 30 second delay isn't great :disappointed:
